envoy/cilium_host_map.cc

```diff
@@ -11,12 +11,124 @@
 namespace Envoy {
 namespace Cilium {
 
-PolicyHostMap::PolicyHostMap(std::unique_ptr<Envoy::Config::Subscription<cilium::NetworkPolicyHosts>>&& subscription,
-				   ThreadLocal::SlotAllocator& tls)
-  : tls_(tls.allocateSlot()), subscription_(std::move(subscription)) {
+template <typename T>
+unsigned int checkPrefix(T addr, unsigned int plen, absl::string_view host) {
+  const unsigned int PLEN_MAX = sizeof(T)*8;
+  if (plen > PLEN_MAX) {
+    throw EnvoyException(fmt::format("NetworkPolicyHosts: Invalid prefix length in \'{}\'", host));
+  }
+  if (plen == 0) {
+    return PLEN_MAX;
+  }
+  // Check for 1-bits after the prefix
+  T mask = (T(1) << (PLEN_MAX - plen)) - 1;
+  if (addr & ntoh(mask)) {
+    throw EnvoyException(fmt::format("NetworkPolicyHosts: Non-prefix bits set in \'{}\'", host));
+  }
+  return plen;
+}
+
+struct ThreadLocalHostMapInitializer : public PolicyHostMap::ThreadLocalHostMap,
+				       public Logger::Loggable<Logger::Id::config> {
+protected:
+  friend class PolicyHostMap; // PolicyHostMap can insert();
+
+  // find the map of the given prefix length, insert in the decreasing order if it does
+  // not exist
+  template <typename M>
+  M& getMap(std::vector<std::pair<unsigned int, M>>& maps, unsigned int plen) {
+    auto it = maps.begin();
+    for (; it != maps.end(); it++) {
+      if (it->first > plen) {
+	ENVOY_LOG(debug, "Skipping map for prefix length {} while looking for {}", it->first, plen);
+	continue; // check the next one
+      }
+      if (it->first == plen) {
+	ENVOY_LOG(debug, "Found existing map for prefix length {}", plen);
+	return it->second;
+      }
+      // Current pair has has smaller prefix, insert before it to maintain order
+      ENVOY_LOG(debug, "Inserting map for prefix length {} before prefix length {}", plen, it->first);
+      break;
+    }
+    // not found, insert before the position 'it'
+    ENVOY_LOG(debug, "Inserting map for prefix length {}", plen);
+    return maps.emplace(it, std::make_pair(plen, M{}))->second;
+  }
+
+  bool insert(uint32_t addr, unsigned int plen, uint64_t policy) {
+    auto pair = getMap(ipv4_to_policy_, plen).emplace(std::make_pair(addr, policy));
+    return pair.second;
+  }
+
+  bool insert(absl::uint128 addr, unsigned int plen, uint64_t policy) {
+    auto pair = getMap(ipv6_to_policy_, plen).emplace(std::make_pair(addr, policy));
+    return pair.second;
+  }
+
+  void insert(const cilium::NetworkPolicyHosts& proto) {
+    uint64_t policy = proto.policy();
+    const auto& hosts = proto.host_addresses();
+    std::string buf;
+
+    for (const auto& host: hosts) {
+      const char *addr = host.c_str();
+      unsigned int plen = 0;
+
+      // Find the prefix length if any
+      const char *slash = strchr(addr, '/');
+      if (slash != nullptr) {
+	const char *pstr = slash + 1;
+	// Must start with a non-zero digit
```

Comment:
Why not just call `atoi(pstr)` or `strtoul(pstr, NULL, 10)` and then check that it's within the acceptable range for the IP version?

Reply:
OK.